core/lib/node/src/task/mod.rs

```diff
@@ -0,0 +1,76 @@
+//! Tasks define the "runnable" concept of the node, e.g. something that can be launched and runs until the node
+//! is stopped.
+//!
+//! Task is normally defined by implementing two traits [`IntoZkSyncTask`], which acts like a constructor, and
+//! [`ZkSyncTask`], which provides an interface for `ZkSyncNode` to implement the task lifecycle.
+
+use futures::future::BoxFuture;
+use zksync_health_check::CheckHealth;
+
+use crate::node::{NodeContext, StopReceiver};
+
+pub mod healtcheck_server;
+pub mod metadata_calculator;
+pub mod prometheus_exporter;
+
+/// Factory that can create a task.
+// Note: This have to be a separate trait, since `ZkSyncTask` has to be object-safe.
+pub trait IntoZkSyncTask: 'static + Send + Sync {
+    /// Config type for the task.
+    /// It is highly recommended for tasks to have dedicated configs, not shared with other tasks and resources.
+    type Config: 'static + Send + Sync;
+
+    /// Creates a new task.
+    /// `NodeContext` provides an interface to the utilities that task may need, e.g. the runtime handle to perform asynchronous
+    /// calls, or access to the resources.
+    fn create(
+        node: &NodeContext<'_>,
+        config: Self::Config,
+    ) -> Result<Box<dyn ZkSyncTask>, TaskInitError>;
+}
+
+/// A task implementation.
+#[async_trait::async_trait]
+pub trait ZkSyncTask: 'static + Send + Sync {
+    /// Gets the healthcheck for the task, if it exists.
```

Comment:
that looks like sth that can be moved to the IntyZkSyncTask::create() result.

Reply:
in fact, given that all these methods are supposed to be called at most once, this could be a struct rather than a trait. Like `struct Task { healthcheck: Option<...>, task: BoxFuture<...>, after_node_shutdown: BoxFuture<...> }`, assuming that we could move StopReceiver to NodeContext.